cleverhans/attacks_tf.py

```diff
@@ -726,3 +726,126 @@ def compare(x, y):
         # return the best solution found
         o_bestl2 = np.array(o_bestl2)
         return o_bestattack
+
+
+def deepfool_batch(sess, x, pred, logits, grads, X, nb_candidate, overshoot,
+                   max_iter, clip_min, clip_max, nb_classes, feed=None):
+    """
+    Applies the deepfool to a batch of inputs
+    :param sess: TF session
+    :param x: The input placeholder
+    :param pred: The model's symbolic output
+    :param logits: The model's unnormalized output tensor (the input to
+                   the softmax layer)
+    :param grads: Symbolic gradients
+    :param X: Numpy array with sample inputs
+    :param nb_candidate: The number of adversarial class candidates
+    :param overshoot: A termination criterion to prevent vanishing updates
+    :param max_iter: Maximum number of iteration for deepfool
+    :param clip_min: Minimum value for components of the example returned
+    :param clip_max: Maximum value for components of the example returned
+    :param nb_classes: Number of model output classes
+    :return: Adversarial examples
+    """
+
+    X_adv = np.zeros(X.shape)
+    X_adv = deepfool_attack(sess, x, pred, logits, grads, X, nb_candidate,
+                            overshoot, max_iter, clip_min, clip_max, feed=feed)
+
+    return np.asarray(X_adv, dtype=np.float32)
+
+
+def deepfool_attack(sess, x, predictions, logits, grads, sample, nb_candidate,
+                    overshoot, max_iter, clip_min, clip_max, feed=None):
+    """
+    TensorFlow implementation of the deepfool.
+    Paper link: see https://arxiv.org/pdf/1511.04599.pdf
+    :param sess: TF session
+    :param x: The input placeholder
+    :param predictions: The model's symbolic output (linear output,
+        pre-softmax)
+    :param logits: The model's unnormalized output tensor (the input to
+                   the softmax layer)
+    :param grads: Symbolic gradients
```

Comment:
Try to define more precisely which gradients are needed here. It is good also to say that people should use `gradient_graph` to produce these gradients.